statsd/uds.go

```diff
@@ -20,33 +25,90 @@ type udsWriter struct {
-func newUDSWriter(addr string, writeTimeout time.Duration) (*udsWriter, error) {
-	udsAddr, err := net.ResolveUnixAddr("unixgram", addr)
-	if err != nil {
-		return nil, err
-	}
+func newUDSWriter(addr string, writeTimeout time.Duration, transport string) (*udsWriter, error) {
 	// Defer connection to first Write
-	writer := &udsWriter{addr: udsAddr, conn: nil, writeTimeout: writeTimeout}
+	writer := &udsWriter{addr: addr, transport: transport, conn: nil, writeTimeout: writeTimeout}
 	return writer, nil
 }
 
+// retryOnWriteErr returns true if we should retry writing after a write error
+func (w *udsWriter) retryOnWriteErr(err error) bool {
+	if err == nil {
+		return true
+	}
+	// Never retry when using unixgram (to preserve the historical behavior)
+	if w.transport == "unixgram" {
+		return false
+	}
+	// Otherwise we retry on timeout because we might have written a partial packet
+	if networkError, ok := err.(net.Error); ok && networkError.Timeout() {
+		return true
+	}
+	return false
+}
+
+func (w *udsWriter) shouldCloseConnection(err error) bool {
+	if err, isNetworkErr := err.(net.Error); err != nil && (!isNetworkErr || !err.Timeout()) {
+		// Statsd server disconnected, retry connecting at next packet
+		return true
+	}
+	return false
+}
+
+// writeFull writes the whole data to the UDS connection
+func (w *udsWriter) writeFull(data []byte, stopIfNoneWritten bool) (int, error) {
+	written := 0
+	for written < len(data) {
+		n, e := w.conn.Write(data)
+		written += n
+
+		// If we haven't written anything and we're supposed to stop if we can't write anything, return the error
+		if written == 0 && stopIfNoneWritten {
+			return written, e
+		}
+
+		// If there's an error, check if it is retryable
+		if e != nil && !w.retryOnWriteErr(e) {
+			return written, e
+		}
+
+		// When using "unix" we need to be able to finish to write partially written packets once we have started.
+		if w.transport == "unix" {
+			w.conn.SetWriteDeadline(time.Time{})
+		}
+	}
+	return written, nil
+}
+
 // Write data to the UDS connection with write timeout and minimal error handling:
 // create the connection if nil, and destroy it if the statsd server has disconnected
 func (w *udsWriter) Write(data []byte) (int, error) {
+	var n int
 	conn, err := w.ensureConnection()
 	if err != nil {
 		return 0, err
 	}
 
+	// When using streams the deadline will only make us drop the packet if we can't write it at all,
+	// once we've started writing we need to finish.
 	conn.SetWriteDeadline(time.Now().Add(w.writeTimeout))
-	n, e := conn.Write(data)
 
-	if err, isNetworkErr := e.(net.Error); err != nil && (!isNetworkErr || !err.Temporary()) {
-		// Statsd server disconnected, retry connecting at next packet
+	// When using streams, we append the length of the packet to the data
+	if w.transport == "unix" {
+		bs := []byte{0, 0, 0, 0}
+		binary.LittleEndian.PutUint32(bs, uint32(len(data)))
+		_, err = w.writeFull(bs, true)
```

Comment:
There is no `if err != nil` error management here, should we do something when we can't write our header?

Reply:
```
		_, err = w.writeFull(bs, true)
		if err == nil {
			n, err = w.writeFull(data, false)
		}
```

There is because if we fail to write the header we skip the payload and move to error management a bit later